File: CotEditor/Sources/CETextView+Commenting.m
/*
 
 CETextView+Commenting.m
 
 CotEditor
 http://coteditor.com
 
 Created by 1024jp on 2016-01-10.
 
 ------------------------------------------------------------------------------
 
 © 2014-2016 1024jp
 
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
 http://www.apache.org/licenses/LICENSE-2.0
 
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 
 */

#import "CETextView.h"
#import "CEDefaults.h"
#import "CESyntaxDictionaryKeys.h"


@implementation CETextView (Commenting)

#pragma mark Action Messages

// ------------------------------------------------------
/// toggle comment state in selection
- (IBAction)toggleComment:(nullable id)sender
// ------------------------------------------------------
{
    if ([self canUncommentRange:[self selectedRange]]) {
        [self uncomment:sender];
    } else {
        [self commentOut:sender];
    }
}


// ------------------------------------------------------
/// comment out selection appending comment delimiters
- (IBAction)commentOut:(nullable id)sender
// ------------------------------------------------------
{
    if (![self blockCommentDelimiters] && ![self inlineCommentDelimiter]) { return; }
    
    // determine comment out target
    NSRange targetRange;
    if (![sender isKindOfClass:[NSScriptCommand class]] &&
        [[NSUserDefaults standardUserDefaults] boolForKey:CEDefaultCommentsAtLineHeadKey])
    {
        targetRange = [[self string] lineRangeForRange:[self selectedRange]];
    } else {
        targetRange = [self selectedRange];
    }
    // remove last return
    if (targetRange.length > 0 && [[self string] characterAtIndex:NSMaxRange(targetRange) - 1] == '\n') {
        targetRange.length--;
    }
    
    NSString *target = [[self string] substringWithRange:targetRange];
    NSString *beginDelimiter, *endDelimiter;
    NSString *spacer = [[NSUserDefaults standardUserDefaults] boolForKey:CEDefaultAppendsCommentSpacerKey] ? @" " : @"";
    NSString *newString;
    NSRange selected;
    NSUInteger addedChars = 0;
    
    // insert delimiters
    if ([self inlineCommentDelimiter]) {
        beginDelimiter = [self inlineCommentDelimiter];
        
        newString = [target stringByReplacingOccurrencesOfString:@"\n"
                                                      withString:[NSString stringWithFormat:@"\n%@%@", beginDelimiter, spacer]
                                                         options:0
                                                           range:NSMakeRange(0, [target length])];
        newString = [@[beginDelimiter, newString] componentsJoinedByString:spacer];
        addedChars = [newString length] - targetRange.length;
        
    } else if ([self blockCommentDelimiters]) {
        beginDelimiter = [self blockCommentDelimiters][CEBeginDelimiterKey];
        endDelimiter = [self blockCommentDelimiters][CEEndDelimiterKey];
        
        newString = [@[beginDelimiter, target, endDelimiter] componentsJoinedByString:spacer];
        addedChars = [beginDelimiter length] + [spacer length];
    }
    
    // selection
    if ([self selectedRange].length > 0) {
        selected = NSMakeRange(targetRange.location, [newString length]);
    } else {
        selected = NSMakeRange([self selectedRange].location + addedChars, 0);
    }
    
    // replace
    [self replaceWithString:newString range:targetRange selectedRange:selected
                 actionName:NSLocalizedString(@"Comment Out", nil)];
}


// ------------------------------------------------------
/// uncomment selection removing comment delimiters
- (IBAction)uncomment:(nullable id)sender
// ------------------------------------------------------
{
    if (![self blockCommentDelimiters] && ![self inlineCommentDelimiter]) { return; }
    
    BOOL hasUncommented = NO;
    
    // determine uncomment target
    NSRange targetRange;
    if (![sender isKindOfClass:[NSScriptCommand class]] &&
        [[NSUserDefaults standardUserDefaults] boolForKey:CEDefaultCommentsAtLineHeadKey])
    {
        targetRange = [[self string] lineRangeForRange:[self selectedRange]];
    } else {
        targetRange = [self selectedRange];
    }
    // remove last return
    if (targetRange.length > 0 && [[self string] characterAtIndex:NSMaxRange(targetRange) - 1] == '\n') {
        targetRange.length--;
    }
    
    NSString *target = [[self string] substringWithRange:targetRange];
    NSString *beginDelimiter, *endDelimiter;
    NSString *spacer = [[NSUserDefaults standardUserDefaults] boolForKey:CEDefaultAppendsCommentSpacerKey] ? @" " : @"";
    NSString *newString;
    NSUInteger removedChars = 0;
    
    // block comment
    if ([self blockCommentDelimiters]) {
        if ([target length] > 0) {
            beginDelimiter = [self blockCommentDelimiters][CEBeginDelimiterKey];
            endDelimiter = [self blockCommentDelimiters][CEEndDelimiterKey];
            
            // remove comment delimiters
            if ([target hasPrefix:beginDelimiter] && [target hasSuffix:endDelimiter]) {
                removedChars = [beginDelimiter length];
                newString = [target substringWithRange:NSMakeRange([beginDelimiter length],
                                                                   [target length] - [beginDelimiter length] - [endDelimiter length])];
                
                if ([spacer length] > 0 && [newString hasPrefix:spacer] && [newString hasSuffix:spacer]) {
                    newString = [newString substringWithRange:NSMakeRange(1, [newString length] - 2)];
                    removedChars++;
                }
                
                hasUncommented = YES;
            }
        }
    }
    
    // inline comment
    beginDelimiter = [self inlineCommentDelimiter];
    if (!hasUncommented && beginDelimiter) {
        
        // remove comment delimiters
        NSArray<NSString *> *lines = [target componentsSeparatedByString:@"\n"];
        NSMutableArray<NSString *> *newLines = [NSMutableArray array];
        for (NSString *line in lines) {
            NSString *newLine = [line copy];
            if ([line hasPrefix:beginDelimiter]) {
                newLine = [line substringFromIndex:[beginDelimiter length]];
                
                if ([spacer length] > 0 && [newLine hasPrefix:spacer]) {
                    newLine = [newLine substringFromIndex:[spacer length]];
                }
                
                hasUncommented = YES;
            }
            
            [newLines addObject:newLine];
            removedChars += [line length] - [newLine length];
        }
        
        newString = [newLines componentsJoinedByString:@"\n"];
    }
    
    if (!hasUncommented) { return; }
    
    // set selection
    NSRange selection;
    if ([self selectedRange].length > 0) {
        selection = NSMakeRange(targetRange.location, [newString length]);
    } else {
        selection = NSMakeRange([self selectedRange].location, 0);
        selection.location -= MIN(MIN(selection.location, selection.location - targetRange.location), removedChars);
    }
    
    [self replaceWithString:newString range:targetRange selectedRange:selection
                 actionName:NSLocalizedString(@"Uncomment", nil)];
}



#pragma mark Semi-Private Methods

// ------------------------------------------------------
/// whether given range can be uncommented
- (BOOL)canUncommentRange:(NSRange)range
// ------------------------------------------------------
{
    if (![self blockCommentDelimiters] && ![self inlineCommentDelimiter]) { return NO; }
    
    // determine comment out target
    if ([[NSUserDefaults standardUserDefaults] boolForKey:CEDefaultCommentsAtLineHeadKey]) {
        range = [[self string] lineRangeForRange:range];
    }
    // remove last return
    if (range.length > 0 && [[self string] characterAtIndex:NSMaxRange(range) - 1] == '\n') {
        range.length--;
    }
    
    NSString *target = [[self string] substringWithRange:range];
    
    if ([target length] == 0) { return NO; }
    
    if ([self blockCommentDelimiters]) {
        if ([target hasPrefix:[self blockCommentDelimiters][CEBeginDelimiterKey]] &&
            [target hasSuffix:[self blockCommentDelimiters][CEEndDelimiterKey]]) {
            return YES;
        }
    }
    
    if ([self inlineCommentDelimiter]) {
        NSArray<NSString *> *lines = [target componentsSeparatedByString:@"\n"];
        NSUInteger commentLineCount = 0;
        for (NSString *line in lines) {
            if ([line hasPrefix:[self inlineCommentDelimiter]]) {
                commentLineCount++;
            }
        }
        
        return commentLineCount == [lines count];
    }
    
    return NO;
}

@end

File: CotEditor/Sources/CETextView+Commenting.swift
/*
 
 CETextView+Commenting.swift
 
 CotEditor
 https://coteditor.com
 
 Created by 1024jp on 2016-01-10.
 
 ------------------------------------------------------------------------------
 
 © 2014-2016 1024jp
 
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
 http://www.apache.org/licenses/LICENSE-2.0
 
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 
 */

import Cocoa

extension CETextView {
    
    // MARK: Options
    
    struct CommentTypes: OptionSet {
        let rawValue: Int
        
        static let inline = CommentTypes(rawValue: 1 << 0)
        static let block  = CommentTypes(rawValue: 1 << 1)
        
        static let both: CommentTypes = [.inline, .block]
    }
    
    
    
    // MARK: Action Messages
    
    /// toggle comment state in selection
    @IBAction func toggleComment(_ sender: AnyObject?) {
        
        if self.canUncomment(range: self.selectedRange(), partly: false) {
            self.uncomment(types: .both, fromLineHead: UserDefaults.standard.bool(forKey: DefaultKey.commentsAtLineHead.rawValue))
        } else {
            self.commentOut(types: .both, fromLineHead: UserDefaults.standard.bool(forKey: DefaultKey.commentsAtLineHead.rawValue))
        }
    }
    
    
    /// comment out selection appending comment delimiters
    @IBAction func commentOut(_ sender: AnyObject?) {
        
        self.commentOut(types: .both, fromLineHead: false)
    }
    
    
    /// comment out selection appending block comment delimiters
    @IBAction func blockCommentOut(_ sender: AnyObject?) {
        
        self.commentOut(types: .block, fromLineHead: false)
    }
    
    
    /// comment out selection appending inline comment delimiters
    @IBAction func inlineCommentOut(_ sender: AnyObject?) {
        
        self.commentOut(types: .inline, fromLineHead: false)
    }
    
    
    /// uncomment selection removing comment delimiters
    @IBAction func uncomment(_ sender: AnyObject?) {
        
        self.uncomment(types: .both, fromLineHead: false)
    }
    
    
    
    // MARK: Public Methods
    
    /// comment out selection appending comment delimiters
    func commentOut(types: CommentTypes, fromLineHead: Bool) {
        
        guard self.blockCommentDelimiters != nil || self.inlineCommentDelimiter != nil else { return }
        
        guard let string = self.string,
            let selectedRange = string.range(from: self.selectedRange()),
            let targetRange = self.commentingRange(fromLineHead: fromLineHead)
            else { return }
        
        let spacer = UserDefaults.standard.bool(forKey: DefaultKey.appendsCommentSpacer.rawValue) ? " " : ""
        var new: (String, NSRange)?
        
        // insert delimiters
        if let delimiter = self.inlineCommentDelimiter, types.contains(.inline) {
            new = string.inlineCommentOut(delimiter: delimiter, spacer: spacer, range: targetRange, selectedRange: selectedRange)
            
        } else if let blockDelimiters = self.blockCommentDelimiters, types.contains(.block) {
            let delimiters = BlockDelimiters(begin: blockDelimiters[CEBeginDelimiterKey]!, end: blockDelimiters[CEEndDelimiterKey]!)
            
            new = string.blockCommentOut(delimiters: delimiters, spacer: spacer, range: targetRange, selectedRange: selectedRange)
        }
        
        guard let (newString, newSelectedRange) = new else { return }
        
        self.replace(with: newString, range: string.nsRange(from: targetRange), selectedRange: newSelectedRange,
                     actionName: NSLocalizedString("Comment Out", comment: "action name"))
        
    }
    
    
    /// uncomment selection removing comment delimiters
    func uncomment(types: CommentTypes, fromLineHead: Bool) {
        
        guard self.blockCommentDelimiters != nil || self.inlineCommentDelimiter != nil else { return }
        
        guard let string = self.string,
            let selectedRange = string.range(from: self.selectedRange()),
            let targetRange = self.commentingRange(fromLineHead: fromLineHead),
            !targetRange.isEmpty else { return }
        
        let spacer = UserDefaults.standard.bool(forKey: DefaultKey.appendsCommentSpacer.rawValue) ? " " : ""
        var new: (String, NSRange)?
        
        if let blockDelimiters = self.blockCommentDelimiters, types.contains(.block) {
            let delimiters = BlockDelimiters(begin: blockDelimiters[CEBeginDelimiterKey]!, end: blockDelimiters[CEEndDelimiterKey]!)
            
            new = string.blockUncomment(delimiters: delimiters, spacer: spacer, range: targetRange, selectedRange: selectedRange)
        }
        if let delimiter = self.inlineCommentDelimiter, types.contains(.inline) && new == nil {
            new = string.inlineUncomment(delimiter: delimiter, spacer: spacer, range: targetRange, selectedRange: selectedRange)
        }
        
        guard let (newString, newSelectedRange) = new else { return }
        
        self.replace(with: newString, range: string.nsRange(from: targetRange), selectedRange: newSelectedRange,
                     actionName: NSLocalizedString("Uncomment", comment: "action name"))
        
    }
    
    
    /// whether given range can be uncommented
    @objc(canUncommentRange:partly:)
    func canUncomment(range: NSRange, partly: Bool) -> Bool {
        
        guard self.blockCommentDelimiters != nil || self.inlineCommentDelimiter != nil else { return false }
        
        guard
            let string = self.string,
            let targetRange = self.commentingRange(fromLineHead: UserDefaults.standard.bool(forKey: DefaultKey.commentsAtLineHead.rawValue)),
            !targetRange.isEmpty else { return false }
        
        let target = string.substring(with: targetRange)
        
        if let blockDelimiters = self.blockCommentDelimiters {
            let delimiters = BlockDelimiters(begin: blockDelimiters[CEBeginDelimiterKey]!, end: blockDelimiters[CEEndDelimiterKey]!)
            
            if target.hasPrefix(delimiters.begin) && target.hasSuffix(delimiters.end) {
                return true
            }
        }
        
        if let delimiter = self.inlineCommentDelimiter {
            let lines = target.components(separatedBy: "\n")
            var commentLineCount = 0
            for line in lines {
                if line.hasPrefix(delimiter) {
                    if partly { return true }
                    commentLineCount += 1
                }
            }
            return commentLineCount == lines.count
        }
        
        return false
    }
    
    
    
    // MARK: Private Methods
    
    /// return commenting target range
    private func commentingRange(fromLineHead: Bool) -> Range<String.Index>? {
        
        guard let string = self.string, let selectedRange = string.range(from: self.selectedRange()) else { return nil }
        
        var targetRange = fromLineHead ? string.lineRange(for: selectedRange) : selectedRange
        
        // remove last return
        if string.characters[string.index(before: targetRange.upperBound)] == "\n" {
            targetRange = targetRange.lowerBound..<string.index(before: targetRange.upperBound)
        }
        
        return targetRange
    }
    
}


private extension String {
    
    /// append inline style comment delimiters in range inserting spacer after delimiters and return commented-out string and new selected range
    func inlineCommentOut(delimiter: String, spacer: String, range: Range<Index>, selectedRange: Range<Index>) -> (String, NSRange) {
        
        let target = self.substring(with: range)
        
        let newString = delimiter + spacer + target.replacingOccurrences(of: "\n", with: "\n" + delimiter + spacer)
        let cursorOffset = newString.characters.count - target.characters.count
        
        let newSelectedRange = self.seletedRange(range: range, selectedRange: selectedRange, newString: newString, offset: cursorOffset)
        
        return (newString, newSelectedRange)
    }
    
    
    /// append block style comment delimiters in range inserting spacer between string and delimiters and return commented-out string and new selected range
    func blockCommentOut(delimiters: BlockDelimiters, spacer: String, range: Range<Index>, selectedRange: Range<Index>) -> (String, NSRange) {
        
        let target = self.substring(with: range)
        
        let newString = delimiters.begin + spacer + target + spacer + delimiters.end
        let cursorOffset = delimiters.begin.characters.count + spacer.characters.count
        
        let newSelectedRange = self.seletedRange(range: range, selectedRange: selectedRange, newString: newString, offset: cursorOffset)
        
        return (newString, newSelectedRange)
    }
    
    
    /// remove inline style comment delimiters in range removing also spacers after delimiter and return uncommented string and new selected range
    func inlineUncomment(delimiter: String, spacer: String, range: Range<Index>, selectedRange: Range<Index>) -> (String, NSRange)? {
        
        let target = self.substring(with: range)
        
        var cursorOffset = 0
        
        let lines = target.components(separatedBy: "\n")
        var newLines = [String]()
        for line in lines {
            var newLine = line
            if line.hasPrefix(delimiter) {
                newLine = line.substring(from: line.index(line.startIndex, offsetBy: delimiter.characters.count))
                cursorOffset -= delimiter.characters.count
                
                if !spacer.isEmpty && newLine.hasPrefix(spacer) {
                    newLine = newLine.substring(from: newLine.index(newLine.startIndex, offsetBy: spacer.characters.count))
                    cursorOffset -= spacer.characters.count
                }
            }
            
            newLines.append(newLine)
        }
        
        guard cursorOffset != 0 else { return nil }
        
        let newString = newLines.joined(separator: "\n")
        let newSelectedRange = self.seletedRange(range: range, selectedRange: selectedRange, newString: newString, offset: cursorOffset)
        
        return (newString, newSelectedRange)
    }
    
    
    /// remove block style comment delimiters in range removing also spacers between string and delimiter and return uncommented string and new selected range
    func blockUncomment(delimiters: BlockDelimiters, spacer: String, range: Range<Index>, selectedRange: Range<Index>) -> (String, NSRange)? {
        
        let target = self.substring(with: range)
        
        guard target.hasPrefix(delimiters.begin) && target.hasSuffix(delimiters.end) else { return nil }
        
        let trimFrom = target.index(target.startIndex, offsetBy: delimiters.begin.characters.count)
        let trimTo = target.index(target.endIndex, offsetBy: -delimiters.end.characters.count)
        var newString = target.substring(with: trimFrom..<trimTo)
        var cursorOffset = delimiters.begin.characters.count
        
        if !spacer.isEmpty && newString.hasPrefix(spacer) && newString.hasSuffix(spacer) {
            let trimFrom = newString.index(newString.startIndex, offsetBy: spacer.characters.count)
            let trimTo = newString.index(newString.endIndex, offsetBy: -spacer.characters.count)
            newString = newString.substring(with: trimFrom..<trimTo)
            cursorOffset -= spacer.characters.count
        }
        
        let newSelectedRange = self.seletedRange(range: range, selectedRange: selectedRange, newString: newString, offset: cursorOffset)
        
        return (newString, newSelectedRange)
    }
    
    
    
    // MARK: Private Methods
    
    /// create selected range after commenting-out/uncommenting
    private func seletedRange(range: Range<Index>, selectedRange: Range<Index>, newString: String, offset: Int) -> NSRange {
        
        if selectedRange.isEmpty {
            let selectedLocation = self.utf16.distance(from: self.utf16.startIndex, to: selectedRange.lowerBound.samePosition(in: self.utf16))
            return NSRange(location: selectedLocation + offset, length: 0)
        } else {
            let targetLocation = self.utf16.distance(from: self.utf16.startIndex, to: range.lowerBound.samePosition(in: self.utf16))
            return NSRange(location: targetLocation, length: newString.utf16.count)
        }
    }
    
}
